NMS-14550: Failures when jaeger tracing is enabled on Core server and Minion

deejgregor opened the pull request:
### All Contributors

* [x] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* **If this code is likely to affect the UI, did you name your branch with `-smoke` in it to trigger smoke tests?**
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): [NMS-14550](https://issues.opennms.org/browse/NMS-14550): Failures when jaeger tracing is enabled on Core server and Minion

cgorantla commented:
LGTM other than those changes in [features/endpoints/grafana/client/pom.xml](https://github.com/OpenNMS/opennms/pull/5281/files/9d338c16bf44c7cfca7ef5033a26870df32cd3bc#diff-b355600f31876c81caf6daa4ce67e5750cce24bca8759016f4130fddb84a99b7)

deejgregor commented:
> LGTM other than those changes in [features/endpoints/grafana/client/pom.xml](https://github.com/OpenNMS/opennms/pull/5281/files/9d338c16bf44c7cfca7ef5033a26870df32cd3bc#diff-b355600f31876c81caf6daa4ce67e5750cce24bca8759016f4130fddb84a99b7)

Those changes are there so the feature will still load in Karaf because the newer versions of these libraries include a few dependencies that we don't need at runtime (afaict, at least):
* com.squareup.okhttp3:okhttp:4.9.3: https://repo1.maven.org/maven2/com/squareup/okhttp3/okhttp/4.9.3/okhttp-4.9.3.pom
* com.squareup.okio:okio:2.8.0: https://repo1.maven.org/maven2/com/squareup/okio/okio/2.8.0/okio-2.8.0.pom

These are bundled in the artifact:
```
            <Embed-Dependency>*;artifactId=okhttp|gson|okio</Embed-Dependency>
            <Embed-Transitive>true</Embed-Transitive>
```
-- https://github.com/OpenNMS/opennms/blob/develop/features/endpoints/grafana/client/pom.xml#L24

Other transitive dependencies:
* kotlin libraries (these are new; I'm confirming whether the kotlin libraries are strictly required for this now)
* various crypto libraries (some of these are already excluded)
* Android libraries (some of these are already excluded)

deejgregor commented:
Turned this back into a draft while I work out the Kotlin thing. I also found an issue with inject (seen when trapd starts):
```
...
Caused by: io.jaegertracing.internal.exceptions.UnsupportedFormatException: Builtin.TEXT_MAP_INJECT
        at io.jaegertracing.internal.JaegerTracer.inject(JaegerTracer.java:218)
        at org.opennms.core.tracing.util.TracingInfoCarrier.updateTracingMetadata(TracingInfoCarrier.java:117)
...
```

deejgregor commented:
@cgorantla I cleaned up those exclusions in the core/tracing/jaeger-osgi/pom.xml file and added Kotlin dependencies properly. I also fixed some issues with inject/extract and added a test for it.

cgorantla commented:
@deejgregor  Last we talked, this not in working state yet on Minion. Should we move this to draft state ?

deejgregor commented:
This is ready for review. I got it working for Horizon and the Minion, and added an end-to-end smoke test for both (sending traces to Jaeger that's brought up in a container with testcontainers and make sure we can see traces that we expect).

RangerRick commented on opennms-container/minion/CONFD_README.md:
```diff
@@ -136,12 +136,12 @@ Config specified will be written to `etc/org.opennms.netmgt.trapd.cfg`.
 system:
     properties:
-        jaeger-agent-host: "<host>"
+        JAEGER_AGENT_HOST: "<host>"
```
is there a way to not have to change this, so existing confd configs wouldn't have to be modified?

deejgregor replied:
AFAICT, any existing configs for Jaeger tracing in the Minion aren't working, because the library expects the form `JAEGER_AGENT_HOST` for system properties, not `jaeger-agent-host`, so they'll need to change anyway to work.

deejgregor replied:
I just confirmed with `opennms/minion:30.0.4`: tracing data **does not** get emitted if you just specify `jaeger-agent-host`:
```
system:
  properties:
    jaeger-agent-host: "host.docker.internal"
```

But the library does get initialized properly (but without a place to export to, since it is expecting `JAEGER_AGENT_HOST` or `JAEGER_ENDPOINT` system properties):
```
2022-12-11T22:24:14,134 | INFO  | Blueprint Extender: 1 | JaegerTracerWrapper              | 333 - org.opennms.core.tracing.jaeger-tracer - 30.0.4 | Jaeger tracer initialized with serviceName = MINION@2ffa1e5a-e64a-4943-a71d-1e8ee29e31d8
```

But, if you specify both, like this:
```
system:
  properties:
    jaeger-agent-host: "host.docker.internal"
    JAEGER_AGENT_HOST: "host.docker.internal"
```

Trace data **does** get exported. And, of course, the tracing library obviously is initialized properly:
```
2022-12-11T22:27:07,787 | INFO  | Blueprint Extender: 3 | JaegerTracerWrapper              | 333 - org.opennms.core.tracing.jaeger-tracer - 30.0.4 | Jaeger tracer initialized with serviceName = MINION@2ffa1e5a-e64a-4943-a71d-1e8ee29e31d8
```

If you just specify the `JAEGER_AGENT_HOST` system property that the tracing library expects:
```
system:
  properties:
    JAEGER_AGENT_HOST: "host.docker.internal"
```

You get nothing because the `confd` templates in 30.0.4 look for `jaeger-agent-host` in order to setup the other bits needed for tracing. There is no log showing the tracing library is initialized.

Bonrob2 reviewed (approved):
docs LGTM.
